src/visitors/sympy_replace_solutions_visitor.hpp

```diff
@@ -0,0 +1,465 @@
+                                               const std::string& var);
+
+
+        /// Emplace back the next \p n_next_statements solutions in \ref statements that is marked
+        /// for updating in \ref tags_
+        size_t emplace_back_next_tagged_statements(ast::StatementVector& new_statements,
+                                                   const size_t n_next_statements);
+
+        /// Emplace back all the statements that are marked for updating in \ref tags_
+        size_t emplace_back_all_tagged_statements(ast::StatementVector& new_statements);
+
+        /**
+         * \brief Tag all the statements that depend on \p var for updating
+         *
+         * This is necessary when an assignment has invalidated this variable
+         */
+        size_t tag_dependant_statements(const std::string& var);
+
+        /// Mark that all the statements need updating (probably unused)
+        void tag_all_statements();
+
+        /**
+         * \brief var -> (depends on) vars
+         *
+         * The statements are assignments. Thus, in general, an assigned variable depends on
+         * the variables in the rhs of the assignment operator. This map keeps track of this
+         * dependency
+         */
+        std::unordered_map<std::string, std::unordered_set<std::string>> dependency_map_;
+
+        /**
+         * \brief var -> (statements that depend on) statements
+         *
+         * This the "reverse" of \ref dependency_map_. Given a certain variable it provides
+         * the statements that depend on it. It is a set because we want to print them in
+         * order and we do not want duplicates. The value is the index in \ref statements_ or \ref
+         * tags_
+         */
+        std::unordered_map<std::string, std::set<size_t>> var2dependants_;
+
+        /**
+         * \brief var -> statement that sets that var
+         *
+         * Given a certain variable we get the statement that sets that variable as index in \ref
+         * statements_ or \ref tags_
+         */
+        std::unordered_map<std::string, size_t> var2statement_;
+
+        /// Vector of statements
+        std::vector<std::shared_ptr<ast::Statement>> statements_;
+
+        /**
+         * \brief Keeps track of what statements need updating
+         *
+         * The elements of this set are the indexes of the \ref statements_ vector that need
+         * updating. It is a set because we need to be able to easily find them by value and we need
+         * them ordered to pick "the next one"
+         */
+        std::set<size_t> tags_;
+
+        /**
+         * \brief Max number of times a statement was printed using an \ref
+         * emplace_all_back_statement command
+         *
+         * This is useful to check if, during updates, a variable was assigned.
+         *
+         * For example:
+         *
+         * \f x' = a \f
+         * \f x = a + 1 \f
+         * \f y' = b \f
+         *
```

Comment:
nitpick : I would prefer `\code{.mod}` block for doxygen documentation.

Reply:
done